### PR TITLE
util functions for DNs

### PR DESCRIPTION
**Pull request by vetinari**

### CODE REVIEW

File: dn_util.go [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+import (
+	enchex "encoding/hex"
+	"errors"
+	"strings"
+)
+
+// When true, uses strings.EqualFold to compare the values of an RDN.
+// This is usually needed as true for most values, set to false when
+// your DNs contain case sensitive values.
+var RDNCompareFold bool = true
```
Comment (liggitt) [inline dn_util.go]: Is it possible to have this option not be a global variable? An Equals impl depending on external global state seems weird to me.

File: dn_util.go [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+		var tv []string
+		for _, av := range r.Attributes {
+			tv = append(tv, strings.ToLower(av.Type)+"="+EscapeValue(av.Value))
+		}
+		rdns = append(rdns, strings.Join(tv, "+"))
+	}
+	return strings.Join(rdns, ",")
+}
+
+func EscapeValue(value string) (escaped string) {
```
Comment (liggitt) [inline dn_util.go]: Needs more specific naming to distinguish it from other escape funcs already in the package. Also needs godoc

**Comment by liggitt** [comment]
Some of these look useful, but I'm wary of a big increase in exported APIs before living with them in day to day use for a little while. Do you have examples of these in use in some of your code?

@stevekuznetsov would some of these made your life easier? Any other feedback?

**Comment by stevekuznetsov** [comment]
We aren't doing a lot of DN manipulations - the only check we currently do is to see if a DN falls under another DN in the tree, for which `strings.HasSuffix` works perfectly. I would only be inclined to use an API like this if the queries or search requests already used or supported the DN type. As is, we're using strings to represent the DN so it would be difficult to justify the overhead of creating DN objects to do these checks when string utility methods have the same function.

**Comment by vetinari** [comment]
I can live with maintaining this in a separate module, though a `CanonicalDN()` would be handy, i.e. 
```go
func CanonicalDN (dn string) (string, error) {
   pdn, err := ParseDN(dn); 
   if err != nil {
       return "", err
   } 
   return pdn.String(), nil
}
```
(with the `String()` from the PR)

Currently there's no example in go. Maybe I can show something from a perl script which syncs the users from an AD to an (Open)LDAP server. In this we're dealing with all kinds of user generated RDNs which may contain anything, including `, `, `/` and `(`, `)`.

**Comment by vetinari** [comment]
now in https://github.com/vetinari/go-ldapdn

**Pull request closed without merging**